qpsolvers/solvers/scs_.py

```diff
@@ -105,40 +98,39 @@ def scs_solve_qp(
 
     Notes
     -----
-    As of SCS 2.1.2, the default convergence tolerance ``eps`` is set to
-    ``1e-5``, resulting in inequality constraints that are violated by more
-    than ``1e-6`` as opposed to ``1e-10`` for other solvers e.g. on the README
-    problem. We lower it to ``1e-7`` and switch ``use_indirect=True`` so that
-    SCS behaves closer to the other solvers on this example.
-
-    All other keyword arguments are forwarded as is to SCS. For instance, you
-    can call ``scs_solve_qp(P, q, G, h, use_indirect=True, normalize=True)``.
-    See the solver documentation for details.
+    Keyword arguments are forwarded as is to SCS. For instance, you can call
+    ``scs_solve_qp(P, q, G, h,normalize=True)``. Solver settings for SCS are
+    described `here <https://www.cvxgrp.org/scs/api/settings.html#settings>`_.
     """
+    if isinstance(P, ndarray):
+        warn_about_sparse_conversion("P")
+        P = sparse.csc_matrix(P)
+    if isinstance(G, ndarray):
+        warn_about_sparse_conversion("G")
+        G = sparse.csc_matrix(G)
+    if isinstance(A, ndarray):
+        warn_about_sparse_conversion("A")
+        A = sparse.csc_matrix(A)
+    kwargs["verbose"] = verbose
+    data = {"P": P, "c": q}
     if initvals is not None:
-        warn("note that warm-start values ignored by this wrapper")
-    c_socp, G_socp, h_socp, dims = convert_to_socp(P, q, G, h)
-    kwargs.update(
-        {"eps": eps, "use_indirect": use_indirect, "verbose": verbose}
-    )
-    if A is not None and b is not None:
-        dims["f"] = A.shape[0]  # number of equality constraints
-        A_socp = sparse.hstack(
-            [A, sparse.csc_matrix((A.shape[0], 1))], format="csc"
-        )
-        A_scs = sparse.vstack([A_socp, G_socp], format="csc")
-        b_scs = hstack([b, h_socp])
-        data = {"A": A_scs, "b": b_scs, "c": c_socp}
-        solution = solve(data, dims, **kwargs)
-    else:
-        data = {"A": G_socp, "b": h_socp, "c": c_socp}
-        solution = solve(data, dims, **kwargs)
-    status_val = solution["info"]["statusVal"]
+        data["x"] = initvals
+    if G is None or h is None:
+        raise ValueError("SCS is not available for unconstrained problems")
+    elif A is not None and b is not None:
+        data["A"] = sparse.vstack([G, A, -A], format="csc")
```

Comment:
Stacking the A matrices like this is not the best way to add equality constraints. SCS supports equality constraints directly using the [zero cone](https://www.cvxgrp.org/scs/api/cones.html).

Something like (the order of the stacking is important):
```
data["A"] = sparse.vstack([A , G], format="csc")
data["b"] = hstack([b, h])
cone = {'z': len(b), 'l': len(h)}
```
This should be much faster.

Reply:
Done in 66193ae22386bce81e112959b95ff5ab07676d9a :ok_hand: